### PR TITLE
[ANF] Add a third nic for ANF scaleout

### PR DESCRIPTION
**Pull request by KimForss**
## Problem
Scaleout on ANF requires 3 NICs
## Solution
Add a third NIC that will be deployed if there is a scale out scenario and ANF is used

## Tests
Add at least two db nodes and add the flag  "use_ANF": true to the json file

## Notes
<Additional comments for the PR>

### CODE REVIEW

File: deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf [hunk 1]
```diff
@@ -78,6 +101,11 @@ locals {
   // Filter the list of databases to only HANA platform entries
   hdb = try(local.hdb_list[0], {})
 
+  //ANF support
+  use_ANF = try(local.hdb.use_ANF, false)
+  //Scalout subnet is needed if ANF is used and there are more than one hana node (or more than two if highly available)
+  scaleout_subnet_needed = local.use_ANF && ((length(local.hdb_vms) > 1 && ! local.hdb_ha) || (length(local.hdb_vms) > 2 && local.hdb_ha))
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf:107]: If local.hdb_vms = 2 and local.hdb_ha = true, according to this logic we do not use ANF? Is that the case?

**Review by rukawata — COMMENTED** [review]
Overall looks fine, a couple of questions:
1. Will there be case where we need ANF for anydb?
1. Can we call it storage nic instead? Since in the article [here](https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/sap/sap-hana-scale-out-standby-netapp-files-suse) it refers 3 subnets as:
  - client  => admin_subnet
  - storage => storage_subnet
  -  hana => db_subnet

File: deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf (hunk 1 above)
Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf:107]: yes - because then it is not scaleout - but HA

Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf:107]: At this time anydb is not (yet) supported on ANF - but it is coming

Comment (rukawata) [reply deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf]: Then if the logic is to have more than one node in one site:
```
dbnode_per_site = length(try(local.hdb.dbnodes, [{}])
scaleout_subnet_needed = local.use_ANF && local.dbnode_per_site > 1
```

Comment (rukawata) [reply deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf]: Do you know the timeline? The reason is that so we can already deploy storage subnet in common_infrustructure, just like admin_subnet and db_subnet.

Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf]: I don't know the timeline but I agree that we can be proactive and put it in common_infrastructure

File: deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf [hunk 2]
```diff
@@ -65,7 +69,16 @@ locals {
...
-  // Define this variable to make it easier when implementing existing kv.
+  // SAP Landscape infrastructure
+  landscape_infrastructure = try(local.landscape_tfstate.landscape_infrastructure, {})
+
+  //SAP vnet
+  vnet_sap_arm_id              = try(local.landscape_tfstate.vnet_sap_arm_id, "")
+  vnet_sap_name                = split("/", local.vnet_sap_arm_id)[8]
+  vnet_sap_resource_group_name = split("/", local.vnet_sap_arm_id)[4]
+  var_vnet_sap                 = try(var.infrastructure.vnets.sap, {})
+
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf:80]: This part does not belong to this PR. Can you check?

File: deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf [hunk 3]
```diff
@@ -78,6 +91,12 @@ locals {
   // Filter the list of databases to only HANA platform entries
   hdb = try(local.hdb_list[0], {})
 
+  //ANF support
+  use_ANF = try(local.hdb.use_ANF, false)
+  //Scalout subnet is needed if ANF is used and there are more than one hana node 
+  dbnode_per_site = length(try(local.hdb.dbnodes, [{}]))
+ storage_subnet_needed = local.use_ANF && local.dbnode_per_site > 1
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf:98]: nit: alignment :)

File: deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf [hunk 4]
```diff
@@ -127,19 +146,22 @@ locals {
...
-    computername = try("${dbnode.name}-0", local.computer_names[idx], local.resource_suffixes.vm)
-    role         = try(dbnode.role, "worker")
-    admin_nic_ip = lookup(dbnode, "admin_nic_ips", [false, false])[0]
-    db_nic_ip    = lookup(dbnode, "db_nic_ips", [false, false])[0]
+    name            = try("${dbnode.name}-0", format("%s%s%s%s", local.prefix, var.naming.separator, local.virtualmachine_names[idx], local.resource_suffixes.vm))
+    computername    = try("${dbnode.name}-0", local.computer_names[idx], local.resource_suffixes.vm)
+    role            = try(dbnode.role, "worker")
+    admin_nic_ip    = lookup(dbnode, "admin_nic_ips", [false, false])[0]
+    db_nic_ip       = lookup(dbnode, "db_nic_ips", [false, false])[0]
+    storage_nic_ip = lookup(dbnode, "storage_nic_ips", [false, false])[0]
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf:154]: nit: alignment.

File: deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf [hunk 5]
```diff
@@ -127,19 +146,22 @@ locals {
...
-      computername = try("${dbnode.name}-1", local.computer_names[idx + local.node_count])
-      role         = try(dbnode.role, "worker")
-      admin_nic_ip = lookup(dbnode, "admin_nic_ips", [false, false])[1]
-      db_nic_ip    = lookup(dbnode, "db_nic_ips", [false, false])[1]
+      name            = try("${dbnode.name}-1", format("%s%s%s%s", local.prefix, var.naming.separator, local.virtualmachine_names[idx + local.node_count], local.resource_suffixes.vm))
+      computername    = try("${dbnode.name}-1", local.computer_names[idx + local.node_count])
+      role            = try(dbnode.role, "worker")
+      admin_nic_ip    = lookup(dbnode, "admin_nic_ips", [false, false])[1]
+      db_nic_ip       = lookup(dbnode, "db_nic_ips", [false, false])[1]
+      storage_nic_ip = lookup(dbnode, "storage_nic_ips", [false, false])[1]
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf:164]: nit: alignment.

File: deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf [hunk 6]
```diff
@@ -183,24 +205,26 @@ locals {
...
   // Subnet IP Offsets
   // Note: First 4 IP addresses in a subnet are reserved by Azure
   hdb_ip_offsets = {
-    hdb_lb       = 4
-    hdb_admin_vm = 10
-    hdb_db_vm    = 10
+    hdb_lb          = 4
+    hdb_admin_vm    = 10
+    hdb_db_vm       = 10
+    hdb_storage_vm = 10
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/hdb_node/variables_local.tf:227]: nit: alignment

File: deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/outputs.tf [hunk 7]
```diff
@@ -41,3 +41,7 @@ output "sid_kv_user" {
 output "sid_kv_prvt" {
   value = local.enable_sid_deployment ? azurerm_key_vault.sid_kv_prvt : null
 }
+
+output "storage_subnet" {
+  value = local.enable_db_deployment && local.storage_subnet_needed ? local.sub_storage_exists ? data.azurerm_subnet.storage[0] : azurerm_subnet.storage[0] : null
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/outputs.tf:46]: Can we organize this condition so it's clearer? eg. with parentheses.

File: deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/variables_local.tf [hunk 8]
```diff
@@ -133,6 +133,12 @@ locals {
   //Enable SID deployment
   enable_sid_deployment = local.enable_db_deployment || local.enable_app_deployment
 
+ //ANF support
+  use_ANF = try(local.db.use_ANF, false)
+  //Scalout subnet is needed if ANF is used and there are more than one hana node 
+  dbnode_per_site = length(try(local.db.dbnodes, [{}]))
+  storage_subnet_needed = local.use_ANF && local.dbnode_per_site > 1
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/variables_local.tf:140]: Since we use enable_admin_subnet for admin, can we name it `enable_storage_subnet` instead?

Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/variables_local.tf:140]: done

**Review by rukawata — APPROVED** [review]
